packages/client/lib/sync/sync.ts

```diff
@@ -248,8 +249,21 @@ export abstract class Synchronizer {
   }
 
   /**
-   * Stop synchronization. Returns a promise that resolves once stopped.
+   *
+   * Processes `NEW_BLOCK` announcement from a peer and inserts into local chain if child of chain tip
+   * @param blockData `NEW_BLOCK` received from peer
    */
+  async handleNewBlock(block: Block) {
+    const chainTip = (await this.chain.getLatestHeader()).hash()
+    // If block parent is current chain tip, insert block into chain
+    if (chainTip.toString('hex') === block.header.parentHash.toString('hex')) {
+      await this.chain.putBlocks([block])
+    } else {
+      // If block is beyond current tip, handle as `NEW_BLOCK_HASHES`
+      this.handleNewBlockHashes([[block.header.hash(), block.header.number]])
+    }
+  }
```

Comment:
At least the `await this.chain.putBlocks([block])` part will only work for full sync and needs to be different, so I guess we need to add this into `FullSynchronizer` and add a slighly different version for the `LightSynchronizer`.

One additional big thing is the rebroadcasting functionality for the `NEW_BLOCK` message, see [Block Propagation](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#block-propagation) section from the devp2p spec:

> Newly-mined blocks must be relayed to all nodes. This happens through block propagation, which is a two step process. When a NewBlock announcement message is received from a peer, the client first verifies the basic header validity of the block, checking whether the proof-of-work value is valid. It then sends the block to a small fraction of connected peers (usually the square root of the total number of peers) using the NewBlock message.
> 
> After the header validity check, the client imports the block into its local chain by executing all transactions contained in the block, computing the block's 'post state'. The block's state-root hash must match the computed post state root. Once the block is fully processed, and considered valid, the client sends a NewBlockHashes message about the block to all peers which it didn't notify earlier. Those peers may request the full block later if they fail to receive it via NewBlock from anyone else.

Maybe you also already give this some thought here and eventually also start to add here to the PR. My first idea is that we do this in a generalized method `handleNewBlock()` in the `Sync` class and the two sub class methods are then both calling in.

I actually had to solve very similar problems with this peer-tracking data structures for the "PooledTxHashes rebroadcasting" ETH service PR #1441 I am just working on, see e.g. the data structure I added in 7e2bb4f7a636d4863beec0feec1bb5bb1bcb3e13 and refined on subsequent commits to map peers (by `PeerId`) to txs send out, and later on clean this data structure on periodic intervals.

Maybe you can directly take some larger - somewhat adopted - parts from this over to not reinvent the wheel here. I guess a lot should be fitting here.

Reply:
>At least the await this.chain.putBlocks([block]) part will only work for full sync and needs to be different, so I guess we need to add this into FullSynchronizer and add a slighly different version for the LightSynchronizer.

On the point about the `LightSynchronizer`, I don't see anything in the LES spec about a `NEW_BLOCK` message.  So, assuming that's the case, should the `handleNewBlock` method just move entirely to `FullSynchronizer`?

I'll take a look at your work on rebroadcasting the message.  I hadn't looked that closely at the spec so didn't realize peers had any obligations around re-broadcasting `NEW_BLOCK` in addition to `NEW_BLOCK_HASHES`.